src/cppnv/EnvReader.cc

```diff
@@ -0,0 +1,813 @@
+  }
+
+  if (result == end_of_stream_value) {
+    return success;
+  }
+  //  trim right side of key
+  while (pair->key->key_index > 0) {
+    if (pair->key->key->at(pair->key->key_index - 1) != ' ') {
+      break;
+    }
+    pair->key->key_index--;
+  }
+  if (!pair->key->has_own_buffer()) {
+    const auto tmp_str = new std::string(pair->key->key_index, '\0');
+
+    tmp_str->replace(0,
+                     pair->key->key_index,
+                     *pair->key->key,
+                     0,
+                     pair->key->key_index);
+
+    pair->key->set_own_buffer(tmp_str);;
+  } else {
+    pair->key->clip_own_buffer(pair->key->key_index);
+  }
+  pair->value->value->clear();
+  const read_result value_result = read_value(file, pair->value);
+  if (value_result == end_of_stream_value) {
+    return end_of_stream_value;
+  }
+  if (value_result == comment_encountered || value_result == success) {
+    if (!pair->value->has_own_buffer()) {
+      const auto tmp_str =
+          new std::string(pair->value->value_index, '\0');
+
+      tmp_str->replace(0,
+                       pair->value->value_index,
+                       *pair->value->value,
+                       0,
+                       pair->value->value_index);
+
+      pair->value->
+            set_own_buffer(tmp_str);
+    } else {
+      pair->value->clip_own_buffer(pair->value->value_index);
+    }
+    remove_unclosed_interpolation(pair->value);
+    return success;
+  }
+  if (value_result == empty) {
+    remove_unclosed_interpolation(pair->value);
+    return empty;
+  }
+  if (value_result == end_of_stream_key) {
+    remove_unclosed_interpolation(pair->value);
+    return end_of_stream_key;
+  }
+
+  remove_unclosed_interpolation(pair->value);
+  return fail;
+}
+
+
+int EnvReader::read_pairs(EnvStream* file, std::vector<EnvPair*>* pairs) {
+  int count = 0;
+  auto buffer = std::string(256, '\0');
+
+  while (true) {
+    buffer.clear();
+    EnvPair* pair = new EnvPair();
+    pair->key = new EnvKey();
+    pair->key->key = &buffer;
+    pair->value = new EnvValue();
+    pair->value->value = &buffer;
+    const read_result result = read_pair(file, pair);
+    if (result == end_of_stream_value) {
+      pairs->push_back(pair);
+      count++;
+      break;
+    }
+    if (result == success) {
+      pairs->push_back(pair);
+      count++;
+      continue;
+    }
+
+    delete pair->key;
+    delete pair->value;
+    delete pair;
+    if (result == comment_encountered || result == fail) {
+      continue;
+    }
+    break;
+  }
+
+  return count;
+}
+
+void EnvReader::delete_pair(const EnvPair* pair) {
+  delete pair->key;
+  delete pair->value;
+  delete pair;
+}
+
+void EnvReader::delete_pairs(const std::vector<EnvPair*>* pairs) {
+  for (const auto env_pair : *pairs) {
+    delete_pair(env_pair);
+  }
+}
+
+int EnvReader::read_pairs(EnvStream* file,
+                          std::map<std::string, EnvPair*>* mapped_pairs) {
+  int count = 0;
```

Comment:
nice implementation :) , but can we get different name for this function?  this quiet confusing when the function name is same but different parameters between

 ` int EnvReader::read_pairs(EnvStream* file, std::vector<EnvPair*>* pairs)`, 
  ` int EnvReader::read_pairs(EnvStream* file, std::map<std::string, EnvPair*>* mapped_pairs)`

Reply:
I dropped support for maps when I realized I'd be copying them into std::map allocating everything all over again. These will be dropped.

Reply:
I removed those functions from this push. I am going to hand go over this one more time tomorrow so that'll give you guys time for any more requests.